### ISSUE TITLE
Loading a checkpoint fails with: Chunk does not reside inside dataset (Dimension on index 2)

### ISSUE THREAD

**Issue by finnolec** [ComputationalRadiationPhysics/picongpu#4570]
Hello, 

I am trying to restart a simulation from a checkpoint and I get this error message when loading the checkpoint:
```
Unhandled exception of type 'St13runtime_error' with message 'Chunk does not reside inside dataset (Dimension on index 2. DS: 18706857984 - Chunk: 18446744071883130880)', terminating
Unhandled exception of type 'St13runtime_error' with message 'Chunk does not reside inside dataset (Dimension on index 2. DS: 18706857984 - Chunk: 18446744072264878080)', terminating
Unhandled exception of type 'St13runtime_error' with message 'Chunk does not reside inside dataset (Dimension on index 2. DS: 18706857984 - Chunk: 18446744072029278208)', terminating
Unhandled exception of type 'St13runtime_error' with message 'Chunk does not reside inside dataset (Dimension on index 2. DS: 18706857984 - Chunk: 18446744072175425536)', terminating
```
times the amount of devices and then it looks like this:
```
srun: error: frontier06390: tasks 905-906,909: Segmentation fault
srun: launch/slurm: _step_signal: Terminating StepId=1321174.1
srun: error: frontier09086: tasks 1410,1413-1414: Segmentation fault
srun: error: frontier03578: tasks 521-522,525-526: Segmentation fault
```

The simulation is 7680 * 7680 * 1920 cells big and it runs on Frontier. My PIConGPU version is: f74735e1b826428ae250ce0d1a24f33c73d11851 with defa63065db74d3cf477f333b42b48319a0bf705 additionally put on top. (`git fetch mainline pull/4566/head:pullrequest4566`)
The output from `bpls` on the checkpoint file looks as follows:
```
$bpls checkpoint_10000.bp/
  float     /data/10000/fields/B/x                                      {7680, 7680, 1920}
  float     /data/10000/fields/B/y                                      {7680, 7680, 1920}
  float     /data/10000/fields/B/z                                      {7680, 7680, 1920}
  float     /data/10000/fields/Convolutional PML B/xy                   {1, 1, 18706857984}
  float     /data/10000/fields/Convolutional PML B/xz                   {1, 1, 18706857984}
  float     /data/10000/fields/Convolutional PML B/yx                   {1, 1, 18706857984}
  float     /data/10000/fields/Convolutional PML B/yz                   {1, 1, 18706857984}
  float     /data/10000/fields/Convolutional PML B/zx                   {1, 1, 18706857984}
  float     /data/10000/fields/Convolutional PML B/zy                   {1, 1, 18706857984}
  float     /data/10000/fields/Convolutional PML E/xy                   {1, 1, 18706857984}
  float     /data/10000/fields/Convolutional PML E/xz                   {1, 1, 18706857984}
  float     /data/10000/fields/Convolutional PML E/yx                   {1, 1, 18706857984}
  float     /data/10000/fields/Convolutional PML E/yz                   {1, 1, 18706857984}
  float     /data/10000/fields/Convolutional PML E/zx                   {1, 1, 18706857984}
  float     /data/10000/fields/Convolutional PML E/zy                   {1, 1, 18706857984}
  float     /data/10000/fields/E/x                                      {7680, 7680, 1920}
  float     /data/10000/fields/E/y                                      {7680, 7680, 1920}
  float     /data/10000/fields/E/z                                      {7680, 7680, 1920}
  char      /data/10000/fields/RNGProvider3XorMin                       {7680, 7680, 46080}
  uint64_t  /data/10000/fields/picongpu_idProvider/nextId               {16, 16, 6}
  uint64_t  /data/10000/fields/picongpu_idProvider/startId              {16, 16, 6}
  float     /data/10000/particles/e/momentum/x                          {91613872330}
  float     /data/10000/particles/e/momentum/y                          {91613872330}
  float     /data/10000/particles/e/momentum/z                          {91613872330}
  uint64_t  /data/10000/particles/e/particlePatches/extent/x            {1536}
  uint64_t  /data/10000/particles/e/particlePatches/extent/y            {1536}
  uint64_t  /data/10000/particles/e/particlePatches/extent/z            {1536}
  uint64_t  /data/10000/particles/e/particlePatches/numParticles        {1536}
  uint64_t  /data/10000/particles/e/particlePatches/numParticlesOffset  {1536}
  uint64_t  /data/10000/particles/e/particlePatches/offset/x            {1536}
  uint64_t  /data/10000/particles/e/particlePatches/offset/y            {1536}
  uint64_t  /data/10000/particles/e/particlePatches/offset/z            {1536}
  float     /data/10000/particles/e/position/x                          {91613872330}
  float     /data/10000/particles/e/position/y                          {91613872330}
  float     /data/10000/particles/e/position/z                          {91613872330}
  int32_t   /data/10000/particles/e/positionOffset/x                    {91613872330}
  int32_t   /data/10000/particles/e/positionOffset/y                    {91613872330}
  int32_t   /data/10000/particles/e/positionOffset/z                    {91613872330}
  float     /data/10000/particles/e/weighting                           {91613872330}
```
I would guess, the error comes somehow from loading the PML fields, because it's the same number. One idea: Am I running into problems, because some indexing with `uint32` does not work anymore? The maximum size of a `uint32` is `4294967295` which is smaller than the total amount of cells in the simulation: 
```
7680 * 7680 * 1920 = 113246208000
7680 * 7680 * 1920 / 4294967295 = 26.36718750613909
```
However, I am a bit confused, where exactly these big numbers come from and what exactly is stored in the `/data/10000/fields/Convolutional PML B/yx` etc. It is also larger than the total amount of cells
```
7680 * 7680 * 1920 / 18706857984 = 6.053726825576996
```
I looked into a different, smaller simulation, where I had a checkpoint lying around and there is also such a big PML-field number in the Adios file.
@PrometheusPi @psychocoderHPC

**Comment by finnolec** [comment]
Using the search function of github I found (probably) the same issue in #4182. Did you find a solution to the problem @wmles?

**Comment by PrometheusPi** [comment]
cc'ing @franzpoeschel

**Comment by pordyna** [comment]
This explains why my checkpoints are so large and take 40min to write

**Comment by PrometheusPi** [comment]
Quick check: the PML data accounts for about 50% extra disk usage. Around 90% of the data are zeros.

**Comment by pordyna** [comment]
@PrometheusPi @psychocoderHPC  @steindev I found the issue. The following function is using global absorber thickness to calculate the local pml field size.
https://github.com/ComputationalRadiationPhysics/picongpu/blob/b2a579b32afafba0a981aa156f586096da764b61/include/picongpu/fields/absorber/pml/Field.tpp#L65-L76
With that, we are setting up PLM fields with a size of the GPU surface with the absorber thickness. We do this on **all** GPUs and not only on GPUs with other boundaries (and even there for more than 1 GPU sims it shouldn't be the complete surface)

There is also this rather complicated constructor that somehow setups indexing within these PML fields. I suppose we need to change something here as well.... Maybe just global thickness to local thickness, but who knows^^.  I don't really understand this indexing. 
https://github.com/ComputationalRadiationPhysics/picongpu/blob/b2a579b32afafba0a981aa156f586096da764b61/include/picongpu/fields/absorber/pml/Field.tpp#L107-L118

From what I have seen in the kernel, it looks like we are using the correct local thicknesses there to determine the position relative to PML. So, my gut feeling would be that we are "just" using unnecessary resources and there is no wrong physics in PML. 

Also, in a very rare case of a local GPU domain smaller in one dimension than the combined PML  global thickness in this direction, we would probably get negative cell numbers and `outerLayerBoxLinearSize` bigger than the local grid size.

Do we have here someone who understands the `OuterLayerBox` indexing, who can have a look at what exactly need to be changed there? 

Also thanks @finnolec for verifying the wrong behavior in `getOuterLayerBoxLinearSize` with printfs.

**Comment by franzpoeschel** [comment]
Note: If I understand this issue correctly, the pathological performance
should only be a problem in HDF5. In HDF5, the amount of data written is
determined by the size of the dataset, while in ADIOS it is determined by
the size of the actually written chunks.

**Comment by finnolec** [comment]
This here is the configuration I used to write the checkpoint @psychocoderHPC. Compression is enabled.
```bash
TBG_ADIOS2_configuration="'{                    \
  \"adios2\": {                                 \
    \"engine\": {                               \
      \"type\": \"bp4\"                         \
      , \"parameters\": {                       \
        \"BufferGrowthFactor\": \"2.0\"         \
        , \"InitialBufferSize\": \"28GB\"       \
        , \"AggregatorRatio\" : \"1\"           \
        , \"Profile\": \"On\"                   \
      }                                         \
    }                                           \
    , \"dataset\": {                            \
      \"operators\": [ {                        \
        \"type\": \"blosc\"                     \
        , \"parameters\": {                     \
          \"clevel\": \"1\"                     \
          , \"compressor\": \"zstd\"            \
          , \"doshuffle\": \"BLOSC_BITSHUFFLE\" \
        }                                       \
      } ]                                       \
    }                                           \
  }                                             \
}'"

TBG_checkpoint="--checkpoint.period 10000:10000 \
                --checkpoint.timePeriod 240 \
                --checkpoint.backend openPMD        \
                --checkpoint.openPMD.ext bp         \
                --checkpoint.openPMD.infix '_%T'    \
                --checkpoint.openPMD.dataPreparationStrategy mappedMemory \
                --checkpoint.openPMD.json !TBG_ADIOS2_configuration"
```

**Comment by pordyna** [comment]
@franzpoeschel but are we not writing all chunks? Btw according to @psychocoderHPC what I found is not a bug, but a missing memory optimization. We do it like that, so it is easier with moving window. Where does exactly @finnolec checkpointing error comes from, is not clear to me.  But, it probably has sth to do with the index being larger than uint32.

**Comment by franzpoeschel** [comment]
Entirely possible, I'm on vacation and only skimmed this on my phone ;)
If the zeros are actually explicitly written by Picongpu, then there's
little that either implementation can do to optimize it

**Comment by steindev** [comment]
@finnolec what is the status here?

**Comment by finnolec** [comment]
This is fixed with #4573, sorry forgot to close
